src/dwarf_parser.cpp

```diff
@@ -1,250 +1,213 @@
+SizedType Dwarf::get_stype(lldb::SBType type, bool resolve_structs)
 {
-  Dwarf_Attribute encoding_attr;
-  Dwarf_Word encoding;
-  dwarf_formudata(
-      dwarf_attr_integrate(&type_die, DW_AT_encoding, &encoding_attr),
-      &encoding);
-  return encoding;
-}
+  if (!type.IsValid())
+    return CreateNone();
 
-SizedType Dwarf::get_stype(Dwarf_Die &type_die, bool resolve_structs) const
-{
-  Dwarf_Die type;
-  dwarf_peel_type(&type_die, &type);
-
-  auto tag = dwarf_tag(&type);
-  auto bit_size = dwarf_hasattr(&type, DW_AT_bit_size)
-                      ? dwarf_bitsize(&type)
-                      : dwarf_bytesize(&type) * 8;
-  switch (tag) {
-    case DW_TAG_base_type: {
-      Dwarf_Word encoding = get_type_encoding(type);
-      switch (encoding) {
-        case DW_ATE_boolean:
-        case DW_ATE_unsigned:
-        case DW_ATE_unsigned_char:
-          return CreateUInt(bit_size);
-        case DW_ATE_signed:
-        case DW_ATE_signed_char:
+  auto bit_size = 8 * type.GetByteSize();
+
+  switch (type.GetTypeClass()) {
+    case lldb::eTypeClassBuiltin: {
+      switch (type.GetBasicType()) {
+        case lldb::eBasicTypeBool:
+        case lldb::eBasicTypeChar:
+        case lldb::eBasicTypeSignedChar:
+        case lldb::eBasicTypeWChar:
+        case lldb::eBasicTypeSignedWChar:
+#if LLVM_VERSION_MAJOR >= 15
+        case lldb::eBasicTypeChar8:
+#endif
+        case lldb::eBasicTypeChar16:
+        case lldb::eBasicTypeChar32:
+        case lldb::eBasicTypeShort:
+        case lldb::eBasicTypeInt:
+        case lldb::eBasicTypeInt128:
+        case lldb::eBasicTypeLong:
+        case lldb::eBasicTypeLongLong:
           return CreateInt(bit_size);
+        case lldb::eBasicTypeUnsignedChar:
+        case lldb::eBasicTypeUnsignedWChar:
+        case lldb::eBasicTypeUnsignedShort:
+        case lldb::eBasicTypeUnsignedInt:
+        case lldb::eBasicTypeUnsignedInt128:
+        case lldb::eBasicTypeUnsignedLong:
+        case lldb::eBasicTypeUnsignedLongLong:
+          return CreateUInt(bit_size);
         default:
           return CreateNone();
       }
+      break;
     }
-    case DW_TAG_enumeration_type:
+    case lldb::eTypeClassEnumeration:
       return CreateUInt(bit_size);
-    case DW_TAG_pointer_type: {
-      if (dwarf_hasattr(&type, DW_AT_type)) {
-        Dwarf_Die inner_type = type_of(type);
+    case lldb::eTypeClassPointer:
+    case lldb::eTypeClassReference: {
+      if (auto inner_type = type.GetPointeeType()) {
         return CreatePointer(get_stype(inner_type, false));
       }
       // void *
       return CreatePointer(CreateNone());
     }
-    case DW_TAG_structure_type:
-    case DW_TAG_union_type: {
-      std::string name = dwarf_diename(&type_die);
-      name = (tag == DW_TAG_structure_type ? "struct " : "union ") + name;
+    case lldb::eTypeClassClass:
+    case lldb::eTypeClassStruct:
+    case lldb::eTypeClassUnion: {
+      std::string name = type.GetDisplayTypeName() ?: "<unknown type>";
```

Comment:
How will this `<unknown type>` appear to a user?

Reply:
I need to add tests for anonymous struct/union/enum (the case you are referring to) and anonymous fields.
For anonymous types, we could print the declaration instead of `<unknown type>`? Like `enum { A, B } foo`.
For anonymous fields, we might need to flatten the type to reproduce the behaviour of a regular compiler.

Reply:
Following our chat, `<unknown type>` (now renamed to `<anonymous type>`) would appear when listing the argument of a function (anonymous `enum` or `union`, but we shouldn't encounter this often...), and when printing a `struct` whose fields have anonymous type.
The goal is to revisit this and to correctly print the representation of anonymous types. I'll open an issue for that.